### PR TITLE
lang: Clearer error when handling not initialized accounts

### LINKED ISSUE #1015: Feature Request: Clearer error when handling `Account` & `ProgramAccount`
For [`Account`](https://github.com/project-serum/anchor/blob/1319dc65d6f8aa791f5121e565ab95129d72795d/lang/src/account.rs#L12) and [`ProgramAccount`](https://github.com/project-serum/anchor/blob/1319dc65d6f8aa791f5121e565ab95129d72795d/lang/src/program_account.rs#L18), before [checking ownership](https://github.com/project-serum/anchor/blob/1319dc65d6f8aa791f5121e565ab95129d72795d/lang/src/program_account.rs#L41-L43) - do a check for existence. An important difference between transferring a non-existent account (belonging to the system program and with `lamports == 0`) an account with really incorrect ownership. 
It is suggested to add an error that the account was expected to be already initialized:
```rust
#[msg("The program expected this account to be already initialized")]
AccountNotInitialized,
```
This is an important clarification, since ownership is not the most intuitive marker of existence.

### PR DESCRIPTION
**Pull request by cyphersnake**
Close #1015 
- Add `ErrorCode::AccountNotInitialized`
- Add check in `Account::try_from` & `Account::try_from_unchecked`

### CODE REVIEW

**Comment by fanatid** [comment]
Since this only more details errors and `ProgramAccount` / `ProgramState` are deprecated I do not think that we need a new `trait`.

**Comment by sadsnake42** [comment]
@fanatid 
I added this trait more for clarification of what I am checking and why, rather than for reuse in deprecated structs. I can clarify this in the comments to the `if` statement, but as for me this is more bad code practice, than rustdoc.

And to be honest, I just moved our internal trait to anchor, since this is how we check if the account is initialized or not. 😅
Add a reaction: :+1:  to this post, if you consider it overkill and I'll remove it in favor of a simple use of `if`.

**Comment by sadsnake42** [comment]
I see CI dropped on some tests, I'll figure it out and fix it a little later
Somewhere the error message does not match (it is easy understandable), but somewhere this error occurred unexpectedly, which is surprising!

File: lang/src/lib.rs [hunk 1]
```diff
@@ -154,6 +154,31 @@ pub trait ToAccountInfo<'info> {
...
+    /// Checks if account initialized by call
+    /// [`CreateAccount`](solana_program::system_instruction::SystemInstruction::CreateAccount)
+    /// or [CreateAccountWithSeed](`solana_program::system_instruction::SystemInstruction::CreateAccountWithSeed`)
+    /// solana system instructions
+    ///
+    /// According to the [documentation](https://docs.solana.com/developing/programming-model/accounts#creating),
+    /// an uninitialized account has 0 [lamports](AccountInfo::lamports) and has a [system program](solana_program::system_program) [owner](AccountInfo::owner)
+    fn initialized(&self) -> bool {
+        let account_info = self.to_account_info();
+        account_info.lamports() == 0 && system_program::ID.eq(account_info.owner)
```
Comment (armaniferrante) [inline lang/src/lib.rs:168]: This is inconsistent with the comment, which defines this as uninitialized.

Additionally, the term `initialized` here is potentially confusing. In Anchor, initialized suggests, data space is allocated, the program is the owner, and an account discriminator is assigned (though there's some minor cleanup we need to do internally to actually make this the case).

Comment (sadsnake42) [reply lang/src/lib.rs]: 1. Agree, fixed.
2. Any ideas? `exists` + `not_exists` or `created` + `not_created`?

Comment (armaniferrante) [reply lang/src/lib.rs]: Probably can remove the trait since it's only used in one non deprecated type.

Comment (sadsnake42) [reply lang/src/lib.rs]: As I said above, I wanted to keep the documentation in rustdoc more, but I do not insist. ✌️

File: lang/src/account.rs [hunk 2]
```diff
@@ -34,6 +37,9 @@ impl<'a, T: AccountSerialize + AccountDeserialize + Owner + Clone> Account<'a, T
     /// possible.
     #[inline(never)]
     pub fn try_from_unchecked(info: &AccountInfo<'a>) -> Result<Account<'a, T>, ProgramError> {
+        if system_program::ID.eq(info.owner) && info.lamports() == 0 {
```
Comment (fanatid) [inline lang/src/account.rs:40]: Why `.eq` instead of `==`?

**Comment by fanatid** [comment]
We need a record about new error in CHANGELOG.md

**Comment by fanatid** [comment]
It would be nice if we will have a test somewhere for a new error.

File: lang/src/account.rs (hunk 2 above)
Comment (sadsnake42) [reply lang/src/account.rs:40]: Between the use of `ref == &something` and `ref.eq(something)`, I prefer the latter, but I will correct but more accepted in the project (judging by the `&T::Owner()`)

File: CHANGELOG.md [hunk 3]
```diff
@@ -13,7 +13,11 @@ incremented for features.
 
 ### Fixes
 
-lang: Add `deprecated` attribute to `ProgramAccount` ([#1014](https://github.com/project-serum/anchor/pull/1014)).
+* lang: Add `deprecated` attribute to `ProgramAccount` ([#1014](https://github.com/project-serum/anchor/pull/1014)).
```
Comment (fanatid) [inline CHANGELOG.md:16]: :thinking:

File: tests/errors/tests/errors.js [hunk 4]
```diff
@@ -128,4 +128,19 @@ describe("errors", () => {
...
   });
+
+  it("Emits a account not initialized error", async () => {
+    try {
+      const tx = await program.rpc.accountNotInitializedError({
+        accounts: {
+          notInitializedAccount: (new anchor.web3.Keypair()).publicKey
+        },
+      });
+      assert.ok(false);
```
Comment (fanatid) [inline tests/errors/tests/errors.js:139]: https://nodejs.org/api/assert.html#assertfailmessage :)

**Comment by fanatid** [comment]
LGTM, thanks for the test!

File: CHANGELOG.md (hunk 3 above)
Comment (sadsnake42) [reply CHANGELOG.md:16]: In the whole CHANGELOG, you don't have single items without a "*" at the beginning. I thought it was a typo and corrected it. Boy Scout Rule 😅

File: tests/errors/tests/errors.js (hunk 4 above)
Comment (sadsnake42) [reply tests/errors/tests/errors.js:139]: Just copy from this one:
https://github.com/project-serum/anchor/blob/2cbd973a6d2d2d4e5661f6a6d2cbdf2ab9a9e753/tests/errors/tests/errors.js#L123
Will fix to `assert.fail`

**Comment by armaniferrante** [comment]
The newly added test fails.

**Comment by sadsnake42** [comment]
> The newly added test fails.

Fixed.

@armaniferrante  Is there a way to immediately run tests with updated clients in JS? Or do I just need to learn how repos CI do it?

**Comment by sadsnake42** [comment]
@armaniferrante Do I still need to improve PR somehow? Let me know please

**Comment by armaniferrante** [comment]
> Is there a way to immediately run tests with updated clients in JS? Or I just need to learn how repos CI do it?

I alias the path to my local CLI build and either yarn link or set the file path to the local ts build in the test. Not clean, but it's easy and gets the job done. Would be awesome to see future changes that handled this more elegantly.

**Pull request merged**